File: SignalShareExtension/utils/ShareAppExtensionContext.m
//
// Copyright 2017 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "ShareAppExtensionContext.h"
#import <SignalMessaging/DebugLogger.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>
#import <SignalUI/UIViewController+OWS.h>

NS_ASSUME_NONNULL_BEGIN

@interface ShareAppExtensionContext ()

@property (nonatomic) UIViewController *rootViewController;

@property (atomic) UIApplicationState reportedApplicationState;

@end

#pragma mark -

@implementation ShareAppExtensionContext

@synthesize mainWindow = _mainWindow;
@synthesize appLaunchTime = _appLaunchTime;
@synthesize appForegroundTime = _appForegroundTime;

- (instancetype)initWithRootViewController:(UIViewController *)rootViewController
{
    self = [super init];

    if (!self) {
        return self;
    }

    OWSAssertDebug(rootViewController);

    _rootViewController = rootViewController;

    self.reportedApplicationState = UIApplicationStateActive;

    NSDate *launchDate = [NSDate new];
    _appLaunchTime = launchDate;
    _appForegroundTime = launchDate;

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(extensionHostDidBecomeActive:)
                                                 name:NSExtensionHostDidBecomeActiveNotification
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(extensionHostWillResignActive:)
                                                 name:NSExtensionHostWillResignActiveNotification
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(extensionHostDidEnterBackground:)
                                                 name:NSExtensionHostDidEnterBackgroundNotification
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(extensionHostWillEnterForeground:)
                                                 name:NSExtensionHostWillEnterForegroundNotification
                                               object:nil];

    return self;
}

#pragma mark - Notifications

- (void)extensionHostDidBecomeActive:(NSNotification *)notification
{
    OWSAssertIsOnMainThread();

    OWSLogInfo(@"");

    self.reportedApplicationState = UIApplicationStateActive;

    [BenchManager benchWithTitle:@"Slow post DidBecomeActive"
                 logIfLongerThan:0.01
                 logInProduction:YES
                           block:^{
                               [NSNotificationCenter.defaultCenter
                                   postNotificationName:OWSApplicationDidBecomeActiveNotification
                                                 object:nil];
                           }];
}

- (void)extensionHostWillResignActive:(NSNotification *)notification
{
    OWSAssertIsOnMainThread();

    self.reportedApplicationState = UIApplicationStateInactive;

    OWSLogInfo(@"");
    OWSLogFlush();

    [BenchManager benchWithTitle:@"Slow post WillResignActive"
                 logIfLongerThan:0.01
                 logInProduction:YES
                           block:^{
                               [NSNotificationCenter.defaultCenter
                                   postNotificationName:OWSApplicationWillResignActiveNotification
                                                 object:nil];
                           }];
}

- (void)extensionHostDidEnterBackground:(NSNotification *)notification
{
    OWSAssertIsOnMainThread();

    OWSLogInfo(@"");
    OWSLogFlush();

    self.reportedApplicationState = UIApplicationStateBackground;

    [BenchManager benchWithTitle:@"Slow post DidEnterBackground"
                 logIfLongerThan:0.01
                 logInProduction:YES
                           block:^{
                               [NSNotificationCenter.defaultCenter
                                   postNotificationName:OWSApplicationDidEnterBackgroundNotification
                                                 object:nil];
                           }];
}

- (void)extensionHostWillEnterForeground:(NSNotification *)notification
{
    OWSAssertIsOnMainThread();

    OWSLogInfo(@"");

    self.reportedApplicationState = UIApplicationStateInactive;

    [BenchManager benchWithTitle:@"Slow post WillEnterForeground"
                 logIfLongerThan:0.01
                 logInProduction:YES
                           block:^{
                               [NSNotificationCenter.defaultCenter
                                   postNotificationName:OWSApplicationWillEnterForegroundNotification
                                                 object:nil];
                           }];
}

#pragma mark -

- (BOOL)isMainApp
{
    return NO;
}

- (BOOL)isMainAppAndActive
{
    return NO;
}

- (BOOL)isNSE
{
    return NO;
}

- (UIApplicationState)mainApplicationStateOnLaunch
{
    OWSFailDebug(@"Not main app.");

    return UIApplicationStateInactive;
}

- (BOOL)isRTL
{
    static BOOL isRTL = NO;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        // Borrowed from PureLayout's AppExtension compatible RTL support.
        // App Extensions may not access -[UIApplication sharedApplication]; fall back to checking the bundle's
        // preferred localization character direction
        isRTL = [NSLocale characterDirectionForLanguage:[[NSBundle mainBundle] preferredLocalizations][0]]
            == NSLocaleLanguageDirectionRightToLeft;
    });
    return isRTL;
}

- (CGFloat)statusBarHeight
{
    return 20;
}

- (BOOL)isInBackground
{
    return self.reportedApplicationState == UIApplicationStateBackground;
}

- (BOOL)isAppForegroundAndActive
{
    return self.reportedApplicationState == UIApplicationStateActive;
}

- (UIBackgroundTaskIdentifier)beginBackgroundTaskWithExpirationHandler:
    (BackgroundTaskExpirationHandler)expirationHandler
{
    return UIBackgroundTaskInvalid;
}

- (void)endBackgroundTask:(UIBackgroundTaskIdentifier)backgroundTaskIdentifier
{
    OWSAssertDebug(backgroundTaskIdentifier == UIBackgroundTaskInvalid);
}

- (void)ensureSleepBlocking:(BOOL)shouldBeBlocking blockingObjectsDescription:(NSString *)blockingObjectsDescription
{
    OWSLogDebug(@"Ignoring request to block sleep.");
}

- (void)setMainAppBadgeNumber:(NSInteger)value
{
    OWSFailDebug(@"");
}

- (nullable UIViewController *)frontmostViewController
{
    OWSAssertDebug(self.rootViewController);

    return [self.rootViewController findFrontmostViewController:YES];
}

- (void)openSystemSettings
{
    return;
}

- (void)openURL:(NSURL *)url completion:(void (^__nullable)(BOOL))completion
{
}

- (BOOL)isRunningTests
{
    // We don't need to distinguish this in the SAE.
    return NO;
}

- (CGRect)frame
{
    return self.rootViewController.view.frame;
}

- (UIInterfaceOrientation)interfaceOrientation
{
    return UIInterfaceOrientationPortrait;
}

- (void)setNetworkActivityIndicatorVisible:(BOOL)value
{
    OWSFailDebug(@"");
}

- (void)runNowOrWhenMainAppIsActive:(AppActiveBlock)block
{
    OWSFailDebug(@"cannot run main app active blocks in share extension.");
}

- (id<SSKKeychainStorage>)keychainStorage
{
    return [SSKDefaultKeychainStorage shared];
}

- (NSString *)appDocumentDirectoryPath
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *documentDirectoryURL =
        [[fileManager URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
    return [documentDirectoryURL path];
}

- (NSString *)appSharedDataDirectoryPath
{
    NSURL *groupContainerDirectoryURL =
        [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:TSConstants.applicationGroup];
    return [groupContainerDirectoryURL path];
}

- (NSString *)appDatabaseBaseDirectoryPath
{
    return self.appSharedDataDirectoryPath;
}

- (NSUserDefaults *)appUserDefaults
{
    return [[NSUserDefaults alloc] initWithSuiteName:TSConstants.applicationGroup];
}

- (BOOL)canPresentNotifications
{
    return NO;
}

- (BOOL)shouldProcessIncomingMessages
{
    return NO;
}

- (BOOL)hasUI
{
    return YES;
}

- (BOOL)hasActiveCall
{
    return NO;
}

- (NSString *)debugLogsDirPath
{
    return DebugLogger.shareExtensionDebugLogsDirPath;
}

@end

NS_ASSUME_NONNULL_END

File: SignalShareExtension/ShareAppExtensionContext.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalServiceKit

// This is _NOT_ a singleton and will be instantiated each time that the SAE is used.
final class ShareAppExtensionContext: NSObject {

    private let rootViewController: UIViewController

    var mainWindow: UIWindow?

    @Atomic private var internalReportedApplicationState: UIApplication.State = .active

    let appLaunchTime = Date()

    let appForegroundTime = Date()

    private var notificationCenterObservers = [NSObjectProtocol]()

    static private let isRTL: Bool = {
        // Borrowed from PureLayout's AppExtension compatible RTL support.
        // App Extensions may not access UIApplication.sharedApplication.
        // Fall back to checking the bundle's preferred localization character direction
        guard let language = Bundle.main.preferredLocalizations.first else { return false }
        return NSLocale.characterDirection(forLanguage: language) == .rightToLeft
    }()

    init(rootViewController: UIViewController) {
        self.rootViewController = rootViewController

        super.init()

        let mainQueue = OperationQueue.main
        notificationCenterObservers.append(NotificationCenter.default.addObserver(
            forName: NSNotification.Name.NSExtensionHostDidBecomeActive,
            object: nil,
            queue: mainQueue) { [weak self] notification in
                Logger.info("")
                self?.internalReportedApplicationState = .active
                BenchManager.bench(
                    title: "Slow post DidBecomeActive",
                    logIfLongerThan: 0.01,
                    logInProduction: true) {
                        NotificationCenter.default.post(name: NSNotification.Name.OWSApplicationDidBecomeActive, object: nil)
                    }
            }
        )
        notificationCenterObservers.append(NotificationCenter.default.addObserver(
            forName: NSNotification.Name.NSExtensionHostWillResignActive,
            object: nil,
            queue: mainQueue) { [weak self] notification in
                Logger.info("")
                self?.internalReportedApplicationState = .inactive
                BenchManager.bench(
                    title: "Slow post WillResignActive",
                    logIfLongerThan: 0.01,
                    logInProduction: true) {
                        NotificationCenter.default.post(name: NSNotification.Name.OWSApplicationWillResignActive, object: nil)
                    }
            }
        )
        notificationCenterObservers.append(NotificationCenter.default.addObserver(
            forName: NSNotification.Name.NSExtensionHostDidEnterBackground,
            object: nil,
            queue: mainQueue) { [weak self] notification in
                Logger.info("")
                self?.internalReportedApplicationState = .background
                BenchManager.bench(
                    title: "Slow post DidEnterBackground",
                    logIfLongerThan: 0.01,
                    logInProduction: true) {
                        NotificationCenter.default.post(name: NSNotification.Name.OWSApplicationDidEnterBackground, object: nil)
                    }
            }
        )
        notificationCenterObservers.append(NotificationCenter.default.addObserver(
            forName: NSNotification.Name.NSExtensionHostWillEnterForeground,
            object: nil,
            queue: mainQueue) { [weak self] notification in
                Logger.info("")
                self?.internalReportedApplicationState = .inactive
                BenchManager.bench(
                    title: "Slow post WillEnterForeground",
                    logIfLongerThan: 0.01,
                    logInProduction: true) {
                        NotificationCenter.default.post(name: NSNotification.Name.OWSApplicationWillEnterForeground, object: nil)
                    }
            }
        )
    }

    deinit {
        notificationCenterObservers.forEach { observer in
            NotificationCenter.default.removeObserver(observer)
        }
    }
}

extension ShareAppExtensionContext: AppContext {

    var isMainApp: Bool { false }

    var isMainAppAndActive: Bool { false }

    var isNSE: Bool { false }

    var isRTL: Bool { Self.isRTL }

    var isRunningTests: Bool { false }

    var frame: CGRect { rootViewController.view.frame }

    var interfaceOrientation: UIInterfaceOrientation { .portrait }

    var reportedApplicationState: UIApplication.State {
        internalReportedApplicationState
    }

    func isInBackground() -> Bool {
        return reportedApplicationState == .background
    }

    func isAppForegroundAndActive() -> Bool {
        return reportedApplicationState == .active
    }

    func beginBackgroundTask(expirationHandler: @escaping BackgroundTaskExpirationHandler) -> UIBackgroundTaskIdentifier {
        return .invalid
    }

    func endBackgroundTask(_ backgroundTaskIdentifier: UIBackgroundTaskIdentifier) {
        owsAssertBeta(backgroundTaskIdentifier == .invalid)
    }

    func ensureSleepBlocking(_ shouldBeBlocking: Bool, blockingObjectsDescription: String) {
        Logger.debug("Ignoring request to block sleep.")
    }

    var statusBarHeight: CGFloat { 20 }

    func frontmostViewController() -> UIViewController? {
        return rootViewController.findFrontmostViewController(ignoringAlerts: true)
    }

    func openSystemSettings() { }

    func open(_ url: URL, completion: ((Bool) -> Void)? = nil) { }

    func runNowOr(whenMainAppIsActive block: @escaping AppActiveBlock) {
        owsFailBeta("Cannot run main app active blocks in share extension.")
    }

    func keychainStorage() -> SignalServiceKit.SSKKeychainStorage {
        SSKDefaultKeychainStorage.shared
    }

    func appDocumentDirectoryPath() -> String {
        guard let documentDirectoryURL = FileManager.default.urls(
            for: .documentDirectory,
            in: .userDomainMask
        ).last
        else {
            owsFail("Could not find documents directory.")
        }
        return documentDirectoryURL.path
    }

    func appSharedDataDirectoryPath() -> String {
        guard let groupContainerDirectoryURL = FileManager.default.containerURL(
            forSecurityApplicationGroupIdentifier: TSConstants.applicationGroup
        )
        else {
            owsFail("Could not find application group directory.")
        }
        return groupContainerDirectoryURL.path
    }

    func appDatabaseBaseDirectoryPath() -> String {
        appSharedDataDirectoryPath()
    }

    func appUserDefaults() -> UserDefaults {
        return UserDefaults(suiteName: TSConstants.applicationGroup)!
    }

    func mainApplicationStateOnLaunch() -> UIApplication.State {
        owsFailBeta("Not main app.")
        return .inactive
    }

    func canPresentNotifications() -> Bool { false }

    var shouldProcessIncomingMessages: Bool { false }

    var hasUI: Bool { true }

    var debugLogsDirPath: String { DebugLogger.shareExtensionDebugLogsDirPath }

    var hasActiveCall: Bool { false }

    func resetAppDataAndExit() {
        owsFailBeta("Not main app.")
    }
}
